// Copyright (c) YugaByte, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
// in compliance with the License.  You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under the License
// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
// or implied.  See the License for the specific language governing permissions and limitations
// under the License.
//

syntax = "proto3";

package yb.tserver;

option java_package = "org.yb.tserver";

import "yb/common/common.proto";
import "yb/common/pgsql_protocol.proto";
import "yb/common/transaction.proto";
import "yb/common/value.proto";
import "yb/common/wire_protocol.proto";
import "yb/master/master_ddl.proto";
import "yb/master/master_types.proto";
import "yb/master/master_client.proto";
import "yb/rpc/lightweight_message.proto";

service PgClientService {
  rpc Heartbeat(PgHeartbeatRequestPB) returns (PgHeartbeatResponsePB);

  rpc AlterDatabase(PgAlterDatabaseRequestPB) returns (PgAlterDatabaseResponsePB);
  rpc AlterTable(PgAlterTableRequestPB) returns (PgAlterTableResponsePB);
  rpc BackfillIndex(PgBackfillIndexRequestPB) returns (PgBackfillIndexResponsePB);
  rpc CreateDatabase(PgCreateDatabaseRequestPB) returns (PgCreateDatabaseResponsePB);
  rpc CreateSequencesDataTable(PgCreateSequencesDataTableRequestPB)
      returns (PgCreateSequencesDataTableResponsePB);
  rpc CreateTable(PgCreateTableRequestPB) returns (PgCreateTableResponsePB);
  rpc CreateTablegroup(PgCreateTablegroupRequestPB) returns (PgCreateTablegroupResponsePB);
  rpc DropDatabase(PgDropDatabaseRequestPB) returns (PgDropDatabaseResponsePB);
  rpc DropTable(PgDropTableRequestPB) returns (PgDropTableResponsePB);
  rpc DropTablegroup(PgDropTablegroupRequestPB) returns (PgDropTablegroupResponsePB);
  rpc FinishTransaction(PgFinishTransactionRequestPB) returns (PgFinishTransactionResponsePB);
  rpc GetCatalogMasterVersion(PgGetCatalogMasterVersionRequestPB)
      returns (PgGetCatalogMasterVersionResponsePB);
  rpc GetTServerUUID (PgGetTServerUUIDRequestPB) returns (PgGetTServerUUIDResponsePB);
  rpc GetDatabaseInfo(PgGetDatabaseInfoRequestPB) returns (PgGetDatabaseInfoResponsePB);
  rpc GetLockStatus(PgGetLockStatusRequestPB) returns (PgGetLockStatusResponsePB);
  rpc IsInitDbDone(PgIsInitDbDoneRequestPB) returns (PgIsInitDbDoneResponsePB);
  rpc ListLiveTabletServers(PgListLiveTabletServersRequestPB)
      returns (PgListLiveTabletServersResponsePB);
  rpc GetIndexBackfillProgress(PgGetIndexBackfillProgressRequestPB)
      returns (PgGetIndexBackfillProgressResponsePB);
  rpc OpenTable(PgOpenTableRequestPB) returns (PgOpenTableResponsePB);
  rpc GetTablePartitionList(PgGetTablePartitionListRequestPB)
      returns (PgGetTablePartitionListResponsePB);
  rpc Perform(PgPerformRequestPB) returns (PgPerformResponsePB) {
    option (yb.rpc.lightweight_method).sides = PROXY;
  };
  rpc ReserveOids(PgReserveOidsRequestPB) returns (PgReserveOidsResponsePB);
  rpc RollbackToSubTransaction(PgRollbackToSubTransactionRequestPB)
      returns (PgRollbackToSubTransactionResponsePB);
  // DEPRECATED
  rpc SetActiveSubTransaction(PgSetActiveSubTransactionRequestPB)
      returns (PgSetActiveSubTransactionResponsePB);
  rpc TabletServerCount(PgTabletServerCountRequestPB) returns (PgTabletServerCountResponsePB);
  rpc TruncateTable(PgTruncateTableRequestPB) returns (PgTruncateTableResponsePB);
  rpc ValidatePlacement(PgValidatePlacementRequestPB)
      returns (PgValidatePlacementResponsePB);
  rpc GetTableDiskSize(PgGetTableDiskSizeRequestPB)
      returns (PgGetTableDiskSizeResponsePB);
  rpc WaitForBackendsCatalogVersion(PgWaitForBackendsCatalogVersionRequestPB)
      returns (PgWaitForBackendsCatalogVersionResponsePB);

  rpc InsertSequenceTuple(PgInsertSequenceTupleRequestPB) returns (PgInsertSequenceTupleResponsePB);
  rpc UpdateSequenceTuple(PgUpdateSequenceTupleRequestPB) returns (PgUpdateSequenceTupleResponsePB);
  rpc FetchSequenceTuple(PgFetchSequenceTupleRequestPB) returns (PgFetchSequenceTupleResponsePB);
  rpc ReadSequenceTuple(PgReadSequenceTupleRequestPB) returns (PgReadSequenceTupleResponsePB);
  rpc DeleteSequenceTuple(PgDeleteSequenceTupleRequestPB) returns (PgDeleteSequenceTupleResponsePB);
  rpc DeleteDBSequences(PgDeleteDBSequencesRequestPB) returns (PgDeleteDBSequencesResponsePB);
  rpc CheckIfPitrActive(PgCheckIfPitrActiveRequestPB) returns (PgCheckIfPitrActiveResponsePB);
  rpc IsObjectPartOfXRepl(PgIsObjectPartOfXReplRequestPB)
      returns (PgIsObjectPartOfXReplResponsePB);
  rpc GetTserverCatalogVersionInfo(PgGetTserverCatalogVersionInfoRequestPB)
      returns (PgGetTserverCatalogVersionInfoResponsePB);
  rpc CancelTransaction(PgCancelTransactionRequestPB) returns (PgCancelTransactionResponsePB);
  rpc ActiveUniverseHistory(PgActiveUniverseHistoryRequestPB)
      returns (PgActiveUniverseHistoryResponsePB);
  rpc TabletIDMetadata(PgTabletIDMetadataRequestPB)
      returns (PgTabletIDMetadataResponsePB);
  rpc TableIDMetadata(PgTableIDMetadataRequestPB)
      returns (PgTableIDMetadataResponsePB);
  rpc GetActiveTransactionList(PgGetActiveTransactionListRequestPB)
      returns (PgGetActiveTransactionListResponsePB);
}

message PgHeartbeatRequestPB {
  uint64 session_id = 1;
}

message PgHeartbeatResponsePB {
  AppStatusPB status = 1;
  uint64 session_id = 2;
  bytes instance_id = 3;
}

message PgObjectIdPB {
  uint32 database_oid = 1;
  uint32 object_oid = 2;
}

message PgAlterDatabaseRequestPB {
  uint64 session_id = 1;
  string database_name = 2;
  uint32 database_oid = 3;
  string new_name = 4;
}

message PgAlterDatabaseResponsePB {
  AppStatusPB status = 1;
}

message PgBackfillIndexRequestPB {
  uint64 session_id = 1;
  PgObjectIdPB table_id = 2;
}

message PgBackfillIndexResponsePB {
  AppStatusPB status = 1;
}

message PgRenameColumnPB {
  string old_name = 1;
  string new_name = 2;
}

message PgRenameTablePB {
  string database_name = 1;
  string table_name = 2;
}

message PgAlterTableRequestPB {
  uint64 session_id = 1;
  bool use_transaction = 2;
  PgObjectIdPB table_id = 3;
  repeated PgCreateColumnPB add_columns = 4;
  repeated PgRenameColumnPB rename_columns = 5;
  repeated string drop_columns = 6;
  PgRenameTablePB rename_table = 7;
  bool increment_schema_version = 8;
}

message PgAlterTableResponsePB {
  AppStatusPB status = 1;
}

message PgCreateDatabaseRequestPB {
  uint64 session_id = 1;
  bool use_transaction = 2;
  string database_name = 3;
  uint32 database_oid = 4;
  uint32 source_database_oid = 5;
  uint32 next_oid = 6;
  bool colocated = 7;
}

message PgCreateDatabaseResponsePB {
  AppStatusPB status = 1;
}

message PgCreateSequencesDataTableRequestPB {
}

message PgCreateSequencesDataTableResponsePB {
  AppStatusPB status = 1;
}

message PgCreateColumnPB {
  string attr_name = 1;
  int32 attr_num = 2;
  int32 attr_ybtype = 3;
  bool is_hash = 4;
  bool is_range = 5;
  int32 sorting_type = 6;
  int32 attr_pgoid = 7;
  QLValuePB attr_missing_val = 8;
}

message PgCreateSplitBoundaryPB {
  repeated QLValuePB values = 1;
}

message PgCreateTableRequestPB {
  uint64 session_id = 1;

  bool use_transaction = 2;
  PgObjectIdPB table_id = 3;
  string database_name = 4;
  string table_name = 5;
  int32 num_tablets = 6;
  bool is_pg_catalog_table = 7;
  bool is_shared_table = 8;
  bool if_not_exist = 9;

  bool is_colocated_via_database = 10;
  PgObjectIdPB tablegroup_oid = 11;
  // Auto-generated by catalog manager if not specified.
  // Only makes sense for colocated tables.
  // Note: proto3 v3.15 includes an "optional" keyword as a syntactic sugar for this oneof wrapper.
  oneof optional_colocation_id {
    uint32 colocation_id = 20;
  }

  PgObjectIdPB tablespace_oid = 12;
  repeated PgCreateColumnPB create_columns = 13;
  repeated PgCreateSplitBoundaryPB split_bounds = 14;

  // Actual for index only.
  PgObjectIdPB base_table_id = 15;
  bool is_unique_index = 16;
  bool skip_index_backfill = 17;

  bool is_matview = 21;
  PgObjectIdPB matview_pg_table_oid = 18;
  string schema_name = 19;
}

message PgCreateTableResponsePB {
  AppStatusPB status = 1;
}

message PgCreateTablegroupRequestPB {
  uint64 session_id = 1;

  string database_name = 2;

  PgObjectIdPB tablegroup_id = 3;
  PgObjectIdPB tablespace_id = 4;
}

message PgCreateTablegroupResponsePB {
  AppStatusPB status = 1;
}

message PgDropDatabaseRequestPB {
  uint64 session_id = 1;
  string database_name = 3;
  uint32 database_oid = 4;
}

message PgDropDatabaseResponsePB {
  AppStatusPB status = 1;
}

message PgDropTableRequestPB {
  uint64 session_id = 1;
  PgObjectIdPB table_id = 2;
  bool index = 3;
}

message PgDropTableResponsePB {
  AppStatusPB status = 1;
  master.TableIdentifierPB indexed_table = 2;
}

message PgDropTablegroupRequestPB {
  uint64 session_id = 1;
  PgObjectIdPB tablegroup_id = 2;
}

message PgDropTablegroupResponsePB {
  AppStatusPB status = 1;
}

message PgGetCatalogMasterVersionRequestPB {
}

message PgGetTServerUUIDRequestPB {
}

message PgGetTServerUUIDResponsePB {
  AppStatusPB status = 1;
  repeated uint64 top_level_node_id = 2;
}

message PgFinishTransactionRequestPB {
  uint64 session_id = 1;
  bool commit = 2;
  bool ddl_mode = 3;
  // This transaction contains changes to the DocDB schema protobufs.
  bool has_docdb_schema_changes = 4;
}

message PgFinishTransactionResponsePB {
  AppStatusPB status = 1;
}

message PgGetCatalogMasterVersionResponsePB {
  AppStatusPB status = 1;
  uint64 version = 2;
}

message PgGetDatabaseInfoRequestPB {
  uint32 oid = 1;
}

message PgGetDatabaseInfoResponsePB {
  AppStatusPB status = 1;

  master.GetNamespaceInfoResponsePB info = 2;
}

message PgIsInitDbDoneRequestPB {
}

message PgIsInitDbDoneResponsePB {
  AppStatusPB status = 1;
  bool done = 2;
}

message PgGetLockStatusRequestPB {
  // The table to scan. If unset, scan all tables
  bytes table_id = 1;

  // If set, only read the locks associated with this txn_id and ignore the fields
  // min_txn_age_ms and max_num_txns.
  bytes transaction_id = 2;

  // Return lock statuses of transactions with age >= 'min_txn_age_ms'
  uint32 min_txn_age_ms = 3;

  // Limit number of old transactions whose lock info is retrieved.
  uint32 max_num_txns = 4;
}

message PgGetLockStatusResponsePB {
  AppStatusPB status = 1;

  message NodeLockStatusResponsePB {
    bytes permanent_uuid = 1;

    repeated TabletLockInfoPB tablet_lock_infos = 2;
  }

  repeated NodeLockStatusResponsePB node_locks = 2;

  message TransactionList {
    repeated bytes transaction_ids = 1;
  }

  // Map of transactions keyed by the host node uuid.
  map<string, TransactionList> transactions_by_node = 3;
}

message PgListLiveTabletServersRequestPB {
  bool primary_only = 1;
}

message PgLiveTabletServerPB {
  string uuid = 1;
  string hostname = 2;
  string placement_uuid = 3;
  string cloud = 4;
  string region = 5;
  string zone = 6;
  bool is_primary = 7;
  string public_ip = 8;
  uint32 pg_port = 9;
}

message PgGetIndexBackfillProgressRequestPB {
 repeated PgObjectIdPB index_ids = 1;
}

message PgGetIndexBackfillProgressResponsePB {
  AppStatusPB status = 1;
  // The ordering of the indexes' results here is expected to be the same as the ordering in the
  // corresponding request PB (PgGetIndexBackfillProgressRequestPB).
  repeated uint64 rows_processed_entries = 2;
}

message PgListLiveTabletServersResponsePB {
  AppStatusPB status = 1;
  repeated PgLiveTabletServerPB servers = 2;
}

message PgOpenTableRequestPB {
  string table_id = 1;
  bool reopen = 2;
  fixed64 invalidate_cache_time_us = 3;
}

message PgTablePartitionsPB {
  uint32 version = 1;
  repeated bytes keys = 2;
}

message PgOpenTableResponsePB {
  AppStatusPB status = 1;

  master.GetTableSchemaResponsePB info = 2;
  PgTablePartitionsPB partitions = 3;
}

message PgGetTablePartitionListRequestPB {
  string table_id = 1;
}

message PgGetTablePartitionListResponsePB {
  AppStatusPB status = 1;
  PgTablePartitionsPB partitions = 2;
}

message PgPerformOpPB {
  oneof op {
    PgsqlWriteRequestPB write = 1;
    PgsqlReadRequestPB read = 2;
  }
  bool read_from_followers = 3;
}

enum ReadTimeManipulation {
  NONE = 0;
  RESTART = 1;
  ENSURE_READ_TIME_IS_SET = 2;
}

// TODO: Remove the OptionalUint32PB helper message after updating protobuf generator to allow
// 'optional' in proto3
message OptionalFixed64PB {
  fixed64 value = 1;
}

message OptionalUint32PB {
  uint32 value = 1;
}

message OptionalUint64PB { uint64 value = 1; }

message PgPerformOptionsPB {
  message CachingInfoPB {
    bytes key = 1;
    OptionalUint32PB lifetime_threshold_ms = 2;
  }

  // Cannot use IsolationLevel enum, since we cannot use proto2 enum in proto3 messages.
  // Replace with IsolationLevel after migration to proto3.
  uint32 isolation = 2;
  bool ddl_mode = 3;
  uint64 txn_serial_no = 4;
  uint64 read_time_serial_no = 20;

  // See ReadHybridTimePB for information about in_txn_limit_ht.
  OptionalFixed64PB in_txn_limit_ht = 5;
  bool use_existing_priority = 6;
  fixed64 priority = 7;
  bool restart_transaction = 8;
  bool defer_read_point = 9;
  ReadTimeManipulation read_time_manipulation = 10;
  ReadHybridTimePB read_time = 11;
  bool use_catalog_session = 12;
  bool force_global_transaction = 13;
  string namespace_id = 14;
  bool use_xcluster_database_consistency = 15;
  uint32 active_sub_transaction_id = 16;
  CachingInfoPB caching_info = 17;
  bool read_from_followers = 18;
  bool trace_requested = 19;
  bool yb_non_ddl_txn_for_sys_tables_allowed = 21;
  AUHMetadataPB auh_metadata = 22;
}

message PgPerformRequestPB {
  uint64 session_id = 1;
  PgPerformOptionsPB options = 2;
  repeated PgPerformOpPB ops = 3;
  fixed64 write_time = 4;
}

message PgPerformResponsePB {
  AppStatusPB status = 1;
  repeated PgsqlResponsePB responses = 2;
  ReadHybridTimePB catalog_read_time = 3;
  fixed64 used_in_txn_limit_ht = 4;
}

message PgReserveOidsRequestPB {
  uint32 database_oid = 1;
  uint32 next_oid = 2;
  uint32 count = 3;
}

message PgReserveOidsResponsePB {
  AppStatusPB status = 1;

  uint32 begin_oid = 2;
  uint32 end_oid = 3;
}

message PgRollbackToSubTransactionRequestPB {
  uint64 session_id = 1;
  uint32 sub_transaction_id = 2;
  PgPerformOptionsPB options = 3;
}

message PgRollbackToSubTransactionResponsePB {
  AppStatusPB status = 1;
}

// DEPRECATED
message PgSetActiveSubTransactionRequestPB {
  uint64 session_id = 1;
  uint32 sub_transaction_id = 2;
  // Used to create transaction when we did not have one yet.
  PgPerformOptionsPB options = 3;
}

// DEPRECATED
message PgSetActiveSubTransactionResponsePB {
  AppStatusPB status = 1;
}

message PgTabletServerCountRequestPB {
  bool primary_only = 1;
}

message PgTabletServerCountResponsePB {
  AppStatusPB status = 1;
  int32 count = 2;
}

message PgTruncateTableRequestPB {
  uint64 session_id = 1;
  PgObjectIdPB table_id = 2;
}

message PgTruncateTableResponsePB {
  AppStatusPB status = 1;
}

message PgPlacementInfoPB {
  string cloud = 1;
  string region = 2;
  string zone = 3;
  int32 min_num_replicas = 4;
  int32 leader_preference = 5;
}

message PgValidatePlacementRequestPB {
  uint64 session_id = 1;
  int32 num_replicas = 2;
  repeated PgPlacementInfoPB placement_infos = 3;
}

message PgValidatePlacementResponsePB {
  AppStatusPB status = 1;
}

message PgInsertSequenceTupleRequestPB {
  uint64 session_id = 1;
  int64 db_oid = 2;
  int64 seq_oid = 3;
  uint64 ysql_catalog_version = 4;
  uint64 ysql_db_catalog_version = 7;
  int64 last_val = 5;
  bool is_called = 6;
}

message PgInsertSequenceTupleResponsePB {
  AppStatusPB status = 1;
}

message PgUpdateSequenceTupleRequestPB {
  uint64 session_id = 1;
  int64 db_oid = 2;
  int64 seq_oid = 3;
  uint64 ysql_catalog_version = 4;
  uint64 ysql_db_catalog_version = 10;
  int64 last_val = 5;
  bool is_called = 6;
  bool has_expected = 7;
  int64 expected_last_val = 8;
  bool expected_is_called = 9;
}

message PgUpdateSequenceTupleResponsePB {
  AppStatusPB status = 1;
  bool skipped = 2;
}

message PgFetchSequenceTupleRequestPB {
  uint64 session_id = 1;
  int64 db_oid = 2;
  int64 seq_oid = 3;
  uint64 ysql_catalog_version = 4;
  uint64 ysql_db_catalog_version = 10;
  // The number of values to fetch. If the cache method is connection, this quantity
  // of values will be returned to the connection. If the cache method is server,
  // this quantity will be retrieved from the master server when it is exhausted, and
  // only one value will be returned to the connection.
  uint32 fetch_count = 5;
  int64 inc_by = 6;
  int64 min_value = 7;
  int64 max_value = 8;
  bool cycle = 9;
}

message PgFetchSequenceTupleResponsePB {
  AppStatusPB status = 1;
  int64 first_value = 2;
  int64 last_value = 3;
}

message PgReadSequenceTupleRequestPB {
  uint64 session_id = 1;
  int64 db_oid = 2;
  int64 seq_oid = 3;
  uint64 ysql_catalog_version = 4;
  uint64 ysql_db_catalog_version = 5;
}

message PgReadSequenceTupleResponsePB {
  AppStatusPB status = 1;
  int64 last_val = 2;
  bool is_called = 3;
}

message PgDeleteSequenceTupleRequestPB {
  uint64 session_id = 1;
  int64 db_oid = 2;
  int64 seq_oid = 3;
}

message PgDeleteSequenceTupleResponsePB {
  AppStatusPB status = 1;
}

message PgDeleteDBSequencesRequestPB {
  uint64 session_id = 1;
  int64 db_oid = 2;
}

message PgDeleteDBSequencesResponsePB {
  AppStatusPB status = 1;
}

message PgGetTableDiskSizeRequestPB {
  PgObjectIdPB table_id = 1;
}

message PgGetTableDiskSizeResponsePB {
  AppStatusPB status = 1;

  // size is 0 for colocated tables
  int64 size = 2;
  int32 num_missing_tablets = 3;
}

message PgCheckIfPitrActiveRequestPB {}

message PgCheckIfPitrActiveResponsePB {
  bool is_pitr_active = 1;
  AppStatusPB status = 2;
}

message PgIsObjectPartOfXReplRequestPB {
  PgObjectIdPB table_id = 2;
}

message PgIsObjectPartOfXReplResponsePB {
  AppStatusPB status = 1;
  bool is_object_part_of_xrepl = 2;
}

message PgGetTserverCatalogVersionInfoRequestPB {
  // True if we only want to return the size of the catalog version map from tserver.
  bool size_only = 1;
  // If > 0, we ask for the catalog version of the given database from the tserver.
  // If == 0, we ask for the catalog version map of all databases from the tserver.
  uint32 db_oid = 2;
}

message PgGetTserverCatalogVersionInfoResponsePB {
  AppStatusPB status = 1;
  repeated TServerDBCatalogVersionInfo entries = 2;
  uint32 num_entries = 3;
}

message PgWaitForBackendsCatalogVersionRequestPB {
  uint64 session_id = 1;
  uint32 database_oid = 2;
  uint64 catalog_version = 3;
}

message PgWaitForBackendsCatalogVersionResponsePB {
  AppStatusPB status = 1;
  // Number of YSQL backends (across all tservers) that have not reached the requested catalog
  // version.  If the field is -1, the counting is in-progress.
  int32 num_lagging_backends = 2;
}

message PgCancelTransactionRequestPB {
  bytes transaction_id = 1;

  // Populating the field is optional. When populated, the cancel request is sent to the specified
  // status tablet alone. If not, the request is forwarded to all tablet servers and is thereafter
  // broadcast to all status tablets hosted by the tablet server.
  bytes status_tablet_id = 2;
}

message PgCancelTransactionResponsePB {
  AppStatusPB status = 1;
}

message PgActiveUniverseHistoryRequestPB {}

message PgActiveUniverseHistoryResponsePB {
  AppStatusPB status = 1;
  repeated WaitStateInfoPB tserver_wait_states = 2;
  repeated WaitStateInfoPB cql_wait_states = 3;
  repeated WaitStateInfoPB bg_wait_states = 4;
}

message PgTabletIDMetadataRequestPB {
  string table_id = 1;
}
message PgTabletIDMetadataResponsePB {
  AppStatusPB status = 1;
  yb.master.TabletLocationsPB tablets = 2;
}

message PgTableIDMetadataRequestPB {}

message PgTableIDMetadataResponsePB {
  AppStatusPB status = 1;
  repeated master.ListTablesResponsePB.TableInfo tables = 2;
}

message PgGetActiveTransactionListRequestPB { OptionalUint64PB session_id = 1; }

message PgGetActiveTransactionListResponsePB {
  message EntryPB {
    uint64 session_id = 1;
    bytes txn_id = 2;
  }
  AppStatusPB status = 1;
  repeated EntryPB entries = 2;
}
